src/compute-client/src/controller/sequential_hydration.rs

```diff
@@ -0,0 +1,262 @@
+                            // new collections.
+                            drop(token);
+                            self.hydrate_collections().await?;
+                        }
+                    }
+                } else {
+                    self.collections.insert(*id, collection);
+                }
+            }
+        }
+        Ok(())
+    }
+
+    /// Allow hydration based on the available capacity.
+    async fn hydrate_collections(&mut self) -> Result<(), anyhow::Error> {
+        let capacity = HYDRATION_CONCURRENCY.get(&self.dyncfg);
+        while self.hydration_count() < capacity {
+            let Some(id) = self.hydration_queue.pop_front() else {
+                // Hydration queue is empty.
+                break;
+            };
+            let Some(collection) = self.collections.get_mut(&id) else {
+                // Collection has already been dropped.
+                continue;
+            };
+
+            debug!(%id, "starting collection hydration");
+            self.client.send(ComputeCommand::Schedule(id)).await?;
+
+            let token = Arc::clone(&self.hydration_token);
+            collection.set_hydrating(token);
+        }
+        Ok(())
+    }
+}
+
+#[async_trait]
+impl<C, T> GenericClient<ComputeCommand<T>, ComputeResponse<T>> for SequentialHydration<C, T>
+where
+    C: ComputeClient<T>,
+    T: Timestamp,
+{
+    async fn send(&mut self, cmd: ComputeCommand<T>) -> Result<(), anyhow::Error> {
+        self.absorb_command(cmd).await
+    }
+
+    async fn recv(&mut self) -> Result<Option<ComputeResponse<T>>, anyhow::Error> {
+        let resp = self.client.recv().await?;
+        if let Some(response) = &resp {
+            self.observe_response(response).await?;
+        }
+        Ok(resp)
+    }
+}
+
+/// Information about a tracked collection.
+#[derive(Debug)]
+struct Collection<T> {
+    /// The as-of frontier at collection creation.
+    as_of: Antichain<T>,
+    /// The current state of the collection.
+    state: State,
+}
+
+impl<T> Collection<T> {
+    /// Create a new `Collection`.
+    fn new(as_of: Antichain<T>) -> Self {
+        Self {
+            as_of,
+            state: State::Created,
+        }
+    }
+
+    /// Advance this collection's state to `Scheduled`.
+    fn set_scheduled(&mut self) {
+        soft_assert_eq_or_log!(self.state, State::Created);
+        self.state = State::Scheduled;
+    }
+
+    fn set_hydrating(&mut self, token: Token) {
+        soft_assert_eq_or_log!(self.state, State::Scheduled);
+        self.state = State::Hydrating(token);
+    }
+}
+
+/// The state of a tracked collection.
+#[derive(Debug, PartialEq, Eq)]
+enum State {
+    /// Collection has been created and is waiting for a `Schedule` command.
+    Created,
+    /// Collection has been scheduled and is waiting for hydration capacity.
+    Scheduled,
+    /// Collection is hydrating and waiting for hydration to complete.
+    Hydrating(Token),
+}
```

Comment:
I was slightly confused by the terminology, and thinking about it I think it's not clear out of which perspective we name the state. For example, `Scheduled` means that we've received a schedule command but haven't passed it on. This means the collection is scheduled by the client holding on to us, but it's not scheduled for the client that we're wrapping. Maybe we could make it clearer by adding a comment to `State` that says it's tracking the state out of the perspective of the client that holds on to us.

Reply:
You are right. Rather than explaining the bad terminology with a comment, maybe we can come up with better terminology? How about `Created`, `QueuedForHydration`, `Hydrating`?

Reply:
I like `QueuedForHydration`, and adding a comment to `State` saying that it describes the state of the tracked collection within the client `C`.

Reply:
I adjusted the variant name and the comment.

> saying that it describes the state of the tracked collection within the client `C`.

I think that might be confusing, since the downstream client `C` doesn't know about the hydration queue. We could further rework the variant names to make it very clear that this is only about the `State` of the `SequentialHydration` client, e.g. `AwaitingSchedule`, `QueuedForHydration`, `AwaitingHydration`. Although I think the current terms are already unambiguous enough.